### PR TITLE
SW-4994 Implemented Vote Controller

### PR DESCRIPTION
**Pull request by tommylau523**
Updated Votes controller to take in final decision. 
Updated store to retrieve user info

### CODE REVIEW

**Comment by tommylau523** [comment]
* **#1762** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/1762?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @tommylau523 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: src/main/kotlin/com/terraformation/backend/accelerator/api/ProjectVotesController.kt [hunk 1]
```diff
@@ -85,13 +101,28 @@ data class VoteSelection(
             "The vote the user has selected. Can be yes/no/conditional or `null` if " +
                 "a vote is not yet selected.")
     val email: String,
-    val voteOption: VoteOption? = null,
-    val conditionalInfo: String? = null,
     val firstName: String? = null,
     val lastName: String? = null,
-)
+    val voteOption: VoteOption? = null,
+    val conditionalInfo: String? = null,
```
Comment (karthikbtf) [inline src/main/kotlin/com/terraformation/backend/accelerator/api/ProjectVotesController.kt]: arrange these alphabetically?

File: src/main/kotlin/com/terraformation/backend/accelerator/api/ProjectVotesController.kt [hunk 2]
```diff
@@ -85,13 +101,28 @@ data class VoteSelection(
...
+) {
+  constructor(
+      model: VoteModel
+  ) : this(
+      userId = model.userId,
+      email = model.email,
+      firstName = model.firstName,
+      lastName = model.lastName,
+      voteOption = model.voteOption,
+      conditionalInfo = model.conditionalInfo,
```
Comment (karthikbtf) [inline src/main/kotlin/com/terraformation/backend/accelerator/api/ProjectVotesController.kt]: assign these in alphabetical order

File: src/main/kotlin/com/terraformation/backend/accelerator/api/ProjectVotesController.kt [hunk 3]
```diff
@@ -85,13 +101,28 @@ data class VoteSelection(
...
+  )
+}
 
-data class PhaseVotes(val phase: CohortPhase, val votes: List<VoteSelection>)
+data class PhaseVotes(
+    val phase: CohortPhase,
+    val decision: VoteOption? = null,
```
Comment (karthikbtf) [inline src/main/kotlin/com/terraformation/backend/accelerator/api/ProjectVotesController.kt:122]: alphabetical order

File: src/main/kotlin/com/terraformation/backend/accelerator/api/ProjectVotesController.kt [hunk 4]
```diff
@@ -24,9 +29,9 @@ import org.springframework.web.bind.annotation.RestController
 @AcceleratorEndpoint
 @RequestMapping("/api/v1/accelerator/projects/{projectId}/votes")
 @RestController
-class ProjectVotesController() {
+class ProjectVotesController(private val voteStore: VoteStore) {
   @ApiResponse200
-  @ApiResponse403
+  @ApiResponse403("attempting to read votes without sufficient privilege")
```
Comment (karthikbtf) [inline src/main/kotlin/com/terraformation/backend/accelerator/api/ProjectVotesController.kt]: "Attempting to read votes without sufficient privilege." ? I see other descriptions in this format, capitalized with period.

File: src/main/kotlin/com/terraformation/backend/accelerator/api/ProjectVotesController.kt [hunk 5]
```diff
@@ -120,15 +151,29 @@ data class ProjectVotesPayload(
...
+                phase = it.key,
+                decision =
+                    decisions
+                        .firstOrNull { decisionModel -> decisionModel.phase == it.key }
+                        ?.decision,
+                it.value.map { model -> VoteSelection(model) },
+            )
+          }
+          .values
+          .toList())
```
Comment (karthikbtf) [inline src/main/kotlin/com/terraformation/backend/accelerator/api/ProjectVotesController.kt]: Nice

File: src/main/kotlin/com/terraformation/backend/accelerator/api/ProjectVotesController.kt [hunk 6]
```diff
@@ -120,15 +151,29 @@ data class ProjectVotesPayload(
...
 data class GetProjectVotesResponsePayload(val votes: ProjectVotesPayload) : SuccessResponsePayload
-
-data class UpsertProjectVotesPayload(
-    val projectId: ProjectId,
-    val phase: CohortPhase,
-    val results: List<UpsertVoteSelection>
-)
-
-data class UpsertProjectVotesResponsePayload(val votes: UpsertProjectVotesPayload) :
-    SuccessResponsePayload
```
Comment (karthikbtf) [inline src/main/kotlin/com/terraformation/backend/accelerator/api/ProjectVotesController.kt:134]: 👍

File: src/main/kotlin/com/terraformation/backend/accelerator/db/VoteStore.kt [hunk 7]
```diff
@@ -27,7 +28,10 @@ class VoteStore(
     requirePermissions { readProjectVotes(projectId) }
     return with(PROJECT_VOTES) {
       dslContext
-          .selectFrom(PROJECT_VOTES)
+          .select(PROJECT_VOTES,asterisk(), USERS.EMAIL, USERS.FIRST_NAME, USERS.LAST_NAME)
```
Comment (karthikbtf) [inline src/main/kotlin/com/terraformation/backend/accelerator/db/VoteStore.kt]: space before asterisk? wondering if spotlessApply catches this

File: src/main/kotlin/com/terraformation/backend/accelerator/model/VoteModel.kt [hunk 8]
```diff
@@ -2,16 +2,19 @@ package com.terraformation.backend.accelerator.model
...
 import com.terraformation.backend.db.default_schema.UserId
+import com.terraformation.backend.db.default_schema.tables.references.USERS
 import org.jooq.Record
 
 data class VoteModel(
     val projectId: ProjectId,
     val phase: CohortPhase,
     val userId: UserId,
+    val email: String,
+    val firstName: String? = null,
```
Comment (karthikbtf) [inline src/main/kotlin/com/terraformation/backend/accelerator/model/VoteModel.kt:14]: alphabetical

**Review by karthikbtf — APPROVED** [review]
Thanks, looks good!
Some comments about how we order properties in models, if my understanding is still good.

**Review by karthikbtf — APPROVED** [review]
Thanks for the changes 👍

**Pull request merged**